File: formation-platform/src/lib/music/pipeline.ts
import { AI_PITCH_SAMPLE_RATE, detectNotesWithAI } from "./ai-pitch";
import { detectDrumHits } from "./drums";
import { detectKey } from "./key";
import { detectTempo } from "./tempo";
import type { Transcription } from "./types";

/**
 * Orchestrates samples -> Transcription. Runs entirely client-side: no
 * audio ever leaves the browser unless the creator explicitly saves the
 * result. `samples` must already be mono, resampled to
 * `AI_PITCH_SAMPLE_RATE` (see decodeToMonoSamples in transcription-studio).
 *
 * Runs both the pitched-note detector (Basic Pitch, for melodies/chords)
 * and the percussion detector (plain DSP, for drums) — they look at
 * different things in the same audio, so a recording with both a melody
 * and a beat gets both back.
 */
export async function transcribeSamples(
  samples: Float32Array,
  onProgress?: (percent: number) => void,
): Promise<Transcription> {
  const notes = await detectNotesWithAI(samples, onProgress);
  const drums = detectDrumHits(samples, AI_PITCH_SAMPLE_RATE);
  const tempo = detectTempo(notes);
  const key = detectKey(notes);
  const durationSec = Math.round((samples.length / AI_PITCH_SAMPLE_RATE) * 1000) / 1000;

  return { tempo, key, durationSec, notes, drums };
}

File: melody-transcriber/backend/app/transcription/pipeline.py
"""Orchestrates audio -> Transcription (the JSON contract in app/schema.py)."""

from __future__ import annotations

from app.audio.loader import load_waveform
from app.schema import Transcription
from app.transcription.key import detect_key
from app.transcription.pitch import detect_notes
from app.transcription.tempo import detect_tempo


def transcribe_file(path: str) -> Transcription:
    y, sr = load_waveform(path)
    duration = round(float(len(y) / sr), 3)

    notes = detect_notes(y, sr)
    tempo = detect_tempo(y, sr)
    key = detect_key(y, sr)

    return Transcription(tempo=tempo, key=key, duration=duration, notes=notes)
